packages/pubsub/src/subscription.js

```diff
@@ -592,157 +442,398 @@ Subscription.prototype.decorateMessage_ = function(message) {
+ *   if (err) {
+ *     // Error handling omitted.
+ *   }
+ * });
  *
  * //-
  * // If the callback is omitted, we'll return a Promise.
  * //-
- * subscription.pull(opts).then(function(data) {
- *   var messages = data[0];
- *   var apiResponse = data[1];
+ * subscription.modifyPushConfig(pushConfig).then(function(data) {
+ *   var apiResponse = data[0];
  * });
  */
-Subscription.prototype.pull = function(options, callback) {
-  var self = this;
-  var MAX_EVENTS_LIMIT = 1000;
-
-  if (!callback) {
-    callback = options;
-    options = {};
-  }
-
-  if (!is.number(options.maxResults)) {
-    options.maxResults = MAX_EVENTS_LIMIT;
+Subscription.prototype.modifyPushConfig = function(config, gaxOpts, callback) {
+  if (is.fn(gaxOpts)) {
+    callback = gaxOpts;
+    gaxOpts = {};
   }
 
-  var protoOpts = {
-    service: 'Subscriber',
-    method: 'pull',
-    timeout: this.timeout
-  };
-
   var reqOpts = {
     subscription: this.name,
-    returnImmediately: !!options.returnImmediately,
-    maxMessages: options.maxResults
+    pushConfig: config
   };
 
-  this.activeRequest_ = this.parent.request(protoOpts, reqOpts, function(err) {
-    self.activeRequest_ = null;
+  this.request({
+    client: 'subscriberClient',
+    method: 'modifyPushConfig',
+    reqOpts: reqOpts,
+    gaxOpts: gaxOpts
+  }, callback);
+};
 
-    var resp = arguments[1];
+/**
+ * Nacks the provided message. If the connection pool is absent, it will be
+ * placed in an internal queue and sent out after 1 second or if the pool is
+ * re-opened before the timeout hits.
+ *
+ * @private
+ *
+ * @param {object} message - The message object.
+ */
+Subscription.prototype.nack_ = function(message) {
+  this.breakLease_(message);
 
+  if (!this.connectionPool) {
+    this.inventory_.nack.push(message.ackId);
+    this.setFlushTimeout_();
+    return;
+  }
+
+  var self = this;
+
+  this.connectionPool.acquire(message.connectionId, function(err, connection) {
     if (err) {
-      if (err.code === 504) {
-        // Simulate a server timeout where no messages were received.
-        resp = {
-          receivedMessages: []
-        };
-      } else {
-        callback(err, null, resp);
-        return;
-      }
+      self.emit('error', err);
+      return;
     }
 
-    var messages = arrify(resp.receivedMessages)
-      .map(function(msg) {
-        return Subscription.formatMessage_(msg, self.encoding);
-      })
-      .map(self.decorateMessage_.bind(self));
+    connection.write({
+      modifyDeadlineAckIds: [message.ackId],
+      modifyDeadlineSeconds: [0]
+    });
+  });
+};
 
-    self.refreshPausedStatus_();
+/**
+ * Opens the ConnectionPool.
+ *
+ * @private
+ */
+Subscription.prototype.openConnection_ = function() {
+  var self = this;
 
-    if (self.autoAck && messages.length !== 0) {
-      var ackIds = messages.map(prop('ackId'));
+  var pool = this.connectionPool = new ConnectionPool(this, {
+    ackDeadline: this.ackDeadline,
```

Comment:
Why not read `ackDeadline` and `maxConnections` from the Subscription instance inside of the ConnectionPool constructor?